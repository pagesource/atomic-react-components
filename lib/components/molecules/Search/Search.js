// @flow
/**
 *
 * Search
 *
 */
import React, { useState, useCallback } from 'react';
import type { ComponentType } from 'react';
import styled from 'styled-components';
import classnames from 'classnames';

import { useSpeechRecognition } from '../../../core/hooks';
import Typeahead from '../Typeahead';
import Button from '../../atoms/Button';
import styles from './Search.style';
import type { Props } from './types';

const Search = ({
  className,
  inputProps,
  dataLists,
  getSelectedResultValue,
  renderListItem,
}: Props) => {
  const [searchString, setSearchString] = useState(inputProps.value || '');
  const [searchLists, setSearchList] = useState();

  const onSearchClick = () => {
    getSelectedResultValue(searchString);
  };

  const customFilter = useCallback(
    (inputText) => {
      const lists = dataLists.filter((name) => {
        return name.toLowerCase().indexOf(inputText.toLowerCase()) !== -1;
      });
      setSearchList(lists);
    },
    [dataLists, setSearchList]
  );

  const onSpeechRecognitionResults = (result) => {
    setSearchString(result);
    customFilter(result);
  };

  const handleSpeechEnd = () => {};

  const { isListening, onSpeechListenHandler, browserSupported } = useSpeechRecognition({
    onSpeechRecognitionResults,
    handleSpeechEnd,
  });

  const onSpeechClick = () => onSpeechListenHandler();

  return (
    <div className={classnames(styles.default, className)}>
      <div className="search-box">
        <Typeahead
          className={classnames('search-txt', inputProps.className)}
          id="typeahead"
          name="wordsearch"
          placeholder="Type any word"
          disabled={false}
          searchValue={searchString}
          onChangeHandler={(e) => {
            setSearchString(e.currentTarget.value);
          }}
          results={searchLists}
          getResults={() => {
            customFilter(searchString);
          }}
          getSelectedResultValue={(value) => {
            setSearchString(value);
            getSelectedResultValue(value);
          }}
          renderCustomResultsView={(item) => renderListItem(item)}
          isFieldLabel
          label="search word"
          minSearchChars={3}
          minItemsInList={3}
        />
        <div className="search-button">
          <Button>
            {browserSupported &&
              (isListening ? (
                <i className="search-icon search-loader" />
              ) : (
                // eslint-disable-next-line jsx-a11y/control-has-associated-label
                <i
                  className="search-icon search-mic"
                  role="button"
                  tabIndex={-1}
                  onClick={onSpeechClick}
                  onKeyPress={onSpeechClick}
                />
              ))}
          </Button>
          <Button onClick={onSearchClick} type="button">
            <i className="search-icon search-btn" />
          </Button>
        </div>
      </div>
    </div>
  );
};

Search.defaultProps = {
  className: '',
  inputProps: {},
};

const StyledSearch: ComponentType<Props> = styled(Search)`
  ${styles};
`;
export default StyledSearch;

export { Search as SearchVanilla };
