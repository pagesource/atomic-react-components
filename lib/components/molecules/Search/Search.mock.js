const defaultConfig = {
  dataLists: [
    'Alligator',
    'Alligation',
    'All',
    'Allout',
    'Bask',
    'Crocodilian',
    'Death Roll',
    'Eggs',
    'Jaws',
    'Reptile',
    'Solitary',
    'Tail',
    'Wetlands',
  ],
  ariaLabel: 'Search',
  results: [
    'Alligator',
    'Alligation',
    'All',
    'Allout',
    'Bask',
    'Crocodilian',
    'Death Roll',
    'Eggs',
    'Jaws',
    'Reptile',
    'Solitary',
    'Tail',
    'Wetlands',
  ],
  getSelectedResultValue: (val) => {
    console.log('final result---', val);
  },
};

export default defaultConfig;
