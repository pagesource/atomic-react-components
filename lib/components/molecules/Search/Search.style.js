import { css } from 'styled-components';

export default css`
  margin: 1rem 1rem;
  .search-box {
    display: flex;
    border-radius: 1.5rem;
    border: var(--border-size-field) solid var(--field-border-primary);
    width: 638px;
    margin: 0 auto;
    height: 50px;
  }

  .search-txt {
    display: inherit;
    flex: auto;
    flex-grow: 2;
    input {
      font-size: 1rem;
      flex-grow: 1;
      border: none;
      background: none;
      outline: none;
      line-height: 2;
    }
  }

  .search-button {
    display: flex;
    overflow: hidden;
  }

  button {
    padding: 0 0.5rem;
    cursor: pointer;
    background: transparent;
  }

  button:hover {
    background-color: transparent;
  }

  .search-icon {
    width: 1.5rem;
    height: 1.5rem;
    display: block;
  }

  .search-mic {
    background: url(images/googlemic.png) 0 0/24px no-repeat;
  }

  .search-loader {
    background: url(images/search-loader.gif) 0 0/24px no-repeat;
    height: 0.5rem;
  }

  .search-btn {
    background: url(images/search.svg) 0 0 no-repeat;
  }

  .search-txt ul {
    position: absolute;
    top: 65px;
    bottom: 0;
    left: 0;
    right: 0;
    overflow: hidden;
    z-index: -1;
    width: 638px;
    margin: 0 auto;
  }
  ${(props) => (props.inheritedStyles ? props.inheritedStyles : '')};
`;
