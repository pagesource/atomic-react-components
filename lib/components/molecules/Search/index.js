// @flow
export { default } from './Search';
export { SearchVanilla } from './Search';
