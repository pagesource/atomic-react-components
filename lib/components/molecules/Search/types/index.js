// @flow
import type { Node } from 'react';

export type Props = {
  children: Node,
  className?: string,
  inputProps: { className?: string, value?: string },
  onSpeechEnd: Function,
  onSpeechStart: Function,
  dataLists: [],
  getSelectedResultValue: Function,
  renderListItem: Function,
};
