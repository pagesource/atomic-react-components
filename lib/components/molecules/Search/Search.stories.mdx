import { Meta, Story, Preview, Props } from '@storybook/addon-docs/blocks';

import defaultConfig from './Search.mock';
import Search, { SearchVanilla } from '.';

const ListItem = ({ children }) => children;

<Meta title="Molecules/Search" component={SearchVanilla} />

### Knobs
<Preview>
  <Story name="Knobs">
  <SearchVanilla
      {...defaultConfig}
      className={`hide-default-sample ${defaultConfig.className}`}
    />
  </Story>
</Preview>

### Default
<Preview>
  <Story name="Default">
  <Search {...defaultConfig} renderListItem={(item) => <ListItem key={item}>{item}</ListItem>} />
  </Story>
</Preview>


## Props ðŸ”§

<Props of={SearchVanilla} />

