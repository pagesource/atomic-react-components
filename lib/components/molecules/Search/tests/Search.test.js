import { shallow } from 'enzyme';
import Search from '../index';

describe('<Search />', () => {
  let SearchComponent = '';
  beforeEach(() => {
    SearchComponent = shallow(<Search>Test</Search>);
  });

  test('should render correctly', () => {
    expect(SearchComponent).toMatchSnapshot();
  });
});
