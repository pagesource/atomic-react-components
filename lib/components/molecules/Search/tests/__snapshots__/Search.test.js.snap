// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search /> should render correctly 1`] = `
<Search
  className="Search__StyledSearch-sc-1f83lm1-0 bLFnzM"
  inputProps={Object {}}
>
  Test
</Search>
`;
